source/slang/slang-ast-type.cpp

```diff
@@ -352,6 +352,60 @@ Type* NativeRefType::getValueType()
     return as<Type>(_getGenericTypeArg(this, 0));
 }
 
+Val* PtrTypeBase::getAddressSpace()
+{
+    return _getGenericTypeArg(this, 1);
+}
+
+void maybePrintAddrSpaceOperand(StringBuilder& out, Val* addrSpaceVal)
+{
+    AddressSpace addrSpace = AddressSpace::Generic;
+
+    if (auto cintVal = as<ConstantIntVal>(addrSpaceVal))
+    {
+        addrSpace = (AddressSpace)(cintVal->getValue());
+    }
+    switch (addrSpace)
+    {
+    case AddressSpace::Generic:
+        break;
+    case AddressSpace::UserPointer:
+        out << toSlice(", user_ptr");
+        break;
+    case AddressSpace::GroupShared:
+        out << toSlice(", groupshared");
```

Comment:
Can we make the string consistent?
`group_shared`, `thread_local`

or `userptr`

Reply:
I changed the logic here so that ptr(T, userpointer) is emitted as "Ptr<T>" and ptr(T, generic) is emitted as "Addr<T>".